File: Python3/0523. Continuous Subarray Sum.py
class Solution:
    def checkSubarraySum(self, nums: List[int], k: int) -> bool:
        prefix = [0]
        mod = dict()
        for i in range(len(nums)):
            n = nums[i]
            x = prefix[-1] + n
            prefix.append(x)
            if((i != 0 and x % k == 0) or (x % k in mod and i - mod[x % k] > 1)):
                return True
            elif(x % k not in mod):
                mod[x % k] = i

        return False

File: C++/0523. Continuous Subarray Sum.cpp
class Solution {
public:
    bool checkSubarraySum(vector<int>& nums, int k) {
        vector<int> prefix = {0};
        map<int, int> dict;
        for(int i = 0; i < nums.size(); ++i) {
            int n = nums[i];
            int x = prefix.back() + n;
            prefix.push_back(x);
            int mod = x % k;
            if((mod == 0 && i != 0) || (dict.count(mod) && i - dict[mod] > 1)) {
                return true;                
            } else if(dict.count(mod) == 0) {
                dict[mod] = i;
            }
        }
        return false;

    }
};
